packages/txservice/src/provider.ts

```diff
@@ -0,0 +1,177 @@
+import { jsonifyError } from "@connext/nxtp-utils";
+import axios from "axios";
+import { BigNumber, Signer, Wallet, providers } from "ethers";
+import PriorityQueue from "p-queue";
+import { BaseLogger } from "pino";
+
+import { TransactionServiceConfig } from "./config";
+import { ChainError } from "./error";
+import { FullTransaction, MinimalTransaction, ProviderConfig } from "./types";
+
+const { JsonRpcProvider, FallbackProvider } = providers;
+
+/// Could use a more encompassing name, e.g. ChainRpcDispatch, etc
+export class ChainRpcProvider {
+  private provider: providers.JsonRpcProvider | providers.FallbackProvider;
+  private config: { gasInitialBumpPercent: number; gasPriceMinimum: BigNumber; rpcProviderMaxRetries: number; };
+  private signer: Signer;
+  private queue: PriorityQueue = new PriorityQueue({ concurrency: 1 });
+
+  public confirmationsRequired: number;
+  public chainId: number;
+
+  constructor(
+    private readonly log: BaseLogger,
+    chainId: number,
+    signer: string | Signer,
+    providerConfigs: ProviderConfig[],
+    config: TransactionServiceConfig,
+  ) {
+    const { gasInitialBumpPercent, gasPriceMinimum, rpcProviderMaxRetries } = config;
+    this.config = { gasInitialBumpPercent, gasPriceMinimum, rpcProviderMaxRetries };
+    this.confirmationsRequired = config.chainConfirmations.get(chainId) ?? config.defaultConfirmationsRequired;
+    this.chainId = chainId;
+    
+    // Register a provider for each url.
+    // Make sure all providers are ready()
+
+    if (providerConfigs.length === 1) {
+      this.provider = new JsonRpcProvider(providerConfigs[0]);
+    } else {
+      const fallbackProviderConfigs = providerConfigs.map((config) => ({
+        provider: new JsonRpcProvider(
+          {
+            url: config.url,
+            user: config.user,
+            password: config.password,
+          }
+        ),
+        priority: config.priority ?? 1,
+        weight: config.weight ?? 1,
+        stallTimeout: config.stallTimeout,
+      }));
+      // TODO: Quorum is set to 1 here, but we may want to reconfigure later. Normally it is half the sum of the weights,
+      // which might be okay in our case, but for now we have a low bar.
```

Comment:
Will this always default to whichever provider returns fastest?

Reply:
yep, assuming all providers have a weight of 1

I could leave it empty but my worry is what if like >half of providers fail, but one provider is working perfectly? Then the fallback provider will fail

We should always keep going so long as one (albeit one *trustworthy*) provider is still working.

Reply:
Should we even include a quorum then? It seems like our architecture is slightly different.

Reply:
The quorum under the hood by default will be half the sum of the weights - meaning, normally speaking, we need at least 1/2 of the providers to agree.

But we want the txservice to keep going as long as 1 provider is working, so we want a quorum of 1.